nemo/collections/common/prompts/formatter.py

```diff
@@ -0,0 +1,294 @@
+            cls._REGISTERED_FORMATTERS[cls.REGISTER_NAME] = cls
+        if "preamble" in cls.TEMPLATE:
+            assert (
+                len(cls.TEMPLATE["preamble"].get("slots", [])) == 0
+            ), f"{ERR} Slots are not allowed for preamble template, but we found: '{cls.TEMPLATE['preamble']}'"
+        for role in cls.get_roles():
+            template = cls.get_template(role)
+            for slot in cls.get_slots(role):
+                assert (
+                    _mangled(slot) in template
+                ), f"{ERR} Slot '{slot}' not found in template '{template}' for role '{role}'"
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def resolve(cls, name: str) -> Type["PromptFormatter"]:
+        if name not in cls._REGISTERED_FORMATTERS:
+            raise RuntimeError(
+                f"Unknown prompt formatter: '{name}' (known formats: {', '.join(cls._REGISTERED_FORMATTERS.keys())})"
+            )
+        return cls._REGISTERED_FORMATTERS[name]
+
+    @classmethod
+    @lru_cache(1)
+    def get_roles(cls) -> list[str]:
+        return list(cls.TEMPLATE.keys())
+
+    @classmethod
+    def get_slots(cls, role: str) -> dict[str, Modality]:
+        # returns a copy to avoid accidential mutation of a global object by the user
+        return cls.TEMPLATE[role].get("slots", {}).copy()
+
+    @classmethod
+    def get_template(cls, role: str) -> str:
+        return cls.TEMPLATE[role]["template"]
+
+    def get_default_dialog_slots(self) -> list[dict]:
+        """
+        Returns a list of dialog turns that can be used as a skeleton to fill with actual slot values.
+        If ``PromptFormatter`` was initialized with ``defaults`` argument, this method will return the
+        defaults. Otherwise, every slot is pre-filled with ``None``.
+        """
+        if self._defaults is not None:
+            return self._defaults.copy()
+        else:
+            return [
+                {"role": role, "slots": {slot: None for slot in self.get_slots(role)}}
+                for role in self.get_roles()
+                if role != self.INFERENCE_ROLE
+            ]
+
+    def encode_turn(
+        self, prompt_template: str, expected_slots: dict[str, Modality], slot_values: dict[str, Any]
+    ) -> list[int]:
+        prompt = prompt_template
+        for slot in expected_slots:
+            # For the final substitution of 'slot' in the template we have to mangle it to '|slot|' anyway,
+            # but 'slot' form enables to use valid python identifiers as **kwargs
+            # for passing slots around in user functions.
+            value = slot_values.get(slot)
+            assert value is not None, f"Missing required {slot=} in {slot_values=} for {prompt_template=}"
+            prompt = prompt.replace(_mangled(slot), value)
+        return self._apply_tokenizer(prompt, lang=slot_values.get(self.PROMPT_LANGUAGE_SLOT))
+
+    def encode_dialog(self, turns: list[dict]) -> dict[str, torch.Tensor]:
```

Comment:
If i understand correctly, this PR is for encoder-decoder models like canary/bestow, where all the (multi-turn) dialogue should be in text. 
can we think little bit about supporting audio modality also in slot values. (may be we should keep audio slots untokenized and replace it with "audio features" later. one way is for prompt formatter to return something like list of lists)

Reply:
+1 to this and Piotr told me he is planning on this as v2. maybe we can resume the discussion at the time

Reply:
sg. 
if possible, lets try to put the skeleton in place, e.g. if slot value needs to re-defined as (value, modality) tuple, return needs to be list of lists/tuples etc.

Reply:
sg. 
Agreed. any audio encoder (especially our 600M ones) has to be run on GPU.